### PR TITLE
feat(drive): handlers error codes

### PR DESCRIPTION
**Pull request by markin-io**
## Issue being fixed or feature implemented

Introduced NotFound, InvalidArgument and Internal ABCI response errors

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [ ] I have made corresponding changes to the documentation

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: packages/dapi/lib/externalApis/drive/ErrorCodes.js [hunk 1]
```diff
@@ -0,0 +1,5 @@
+module.exports = {
+  INTERNAL: 1,
+  INVALID_ARGUMENT: 2,
+  NOT_FOUND: 3,
```
Comment (shumkov) [inline packages/dapi/lib/externalApis/drive/ErrorCodes.js:4]: Unknown error 4 is missed? Why don't to reuse all gRPC error code? They all make sense to me in long run.

File: packages/rs-drive-abci/src/abci/handlers.rs [hunk 2]
```diff
@@ -686,22 +682,33 @@ where
         } else {
             let error = result.errors.first();
 
-            let error_message = if let Some(error) = error {
-                error.to_string()
+            let (code, error_message) = if let Some(error) = error {
```
Comment (shumkov) [inline packages/rs-drive-abci/src/abci/handlers.rs:685]: I would recommend to use From trait

File: packages/rs-drive-abci/src/abci/handlers.rs [hunk 3]
```diff
@@ -686,22 +682,33 @@ where
...
+                    QueryError::NotFound(message) => {
+                        (AbciErrorCodes::NotFound as u32, message.to_owned())
+                    }
+                    QueryError::InvalidArgument(message) => {
+                        (AbciErrorCodes::InvalidArgument as u32, message.to_owned())
+                    }
+                    QueryError::Query(error) => {
+                        (AbciErrorCodes::InvalidArgument as u32, error.to_string())
+                    }
+                    _ => (AbciErrorCodes::Unknown as u32, error.to_string()),
```
Comment (shumkov) [inline packages/rs-drive-abci/src/abci/handlers.rs:696]: This must return en error here because we put something into validation result that we can't handle (programming error)

File: packages/rs-drive-abci/src/query/v0/mod.rs [hunk 4]
```diff
@@ -783,13 +851,21 @@ impl<C> Platform<C> {
...
-                        .transpose());
+                        .ok_or_else(|| {
+                            QueryError::NotFound(format!(
+                                "identity {} not found",
+                                hex::encode(public_key_hash)
+                            ))
+                        })
+                        .and_then(|identity| identity
+                            .serialize_consume_to_bytes()
+                            .map_err(QueryError::Protocol)));
```
Comment (shumkov) [inline packages/rs-drive-abci/src/query/v0/mod.rs:862]: Not sure that we need return basic errors as validation errors

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/js-dash-sdk/src/SDK/Client/Platform/Fetcher/Fetcher.ts [hunk 5]
```diff
@@ -96,13 +96,7 @@ class Fetcher {
...
     // Define query
     const query = async (): Promise<GetIdentityResponse> => {
-      const result = await this.dapiClient.platform
-        .getIdentity(id);
-
-      // TODO(rs-drive-abci): Remove this when rs-drive-abci returns error instead of empty bytes
-      if (result.getIdentity().length === 0) {
-        throw new NotFoundError(`Identity with id "${id}" not found`);
-      }
+      const result = this.dapiClient.platform.getIdentity(id);
```
[the remark this answers is not in the document]
Comment (markin-io) [reply packages/js-dash-sdk/src/SDK/Client/Platform/Fetcher/Fetcher.ts:99]: It reformats function in one line which conflicts with eslint rules. 
So, it's either one line with `eslint-disable` or that :)

File: packages/dapi/lib/grpcServer/handlers/createGrpcErrorFromDriveResponse.js [hunk 6]
```diff
@@ -64,7 +56,7 @@ async function createGrpcErrorFromDriveResponse(code, info) {
   const data = decodedInfo.data || {};
 
   // gRPC error codes
```
[the remark this answers is not in the document]
Comment (markin-io) [reply packages/dapi/lib/grpcServer/handlers/createGrpcErrorFromDriveResponse.js]: FIxed

File: packages/js-dash-sdk/src/SDK/Client/Platform/Fetcher/Fetcher.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (markin-io) [reply packages/js-dash-sdk/src/SDK/Client/Platform/Fetcher/Fetcher.ts]: Fixed

File: packages/js-dash-sdk/src/SDK/Client/Platform/Fetcher/Fetcher.ts [hunk 7]
```diff
@@ -123,10 +117,6 @@ class Fetcher {
       const result = await this.dapiClient.platform
```
[the remark this answers is not in the document]
Comment (markin-io) [reply packages/js-dash-sdk/src/SDK/Client/Platform/Fetcher/Fetcher.ts]: Fixed

File: packages/rs-drive-abci/src/error/handlers.rs [hunk 8]
```diff
@@ -0,0 +1,127 @@
+use crate::error::query::QueryError;
+
+/// ABCI handlers errors
+#[derive(Debug, thiserror::Error)]
+pub enum HandlerError {
+    /// ABCI Handler error (Cancelled)
+    #[error("ABCI Handler error (Cancelled): {0}")]
```
Comment (shumkov) [inline packages/rs-drive-abci/src/error/handlers.rs:7]: Do you think we need all this prefix?

File: packages/rs-drive-abci/src/query/v0/mod.rs [hunk 9]
```diff
@@ -71,20 +71,20 @@ fn convert_key_request_type(
             let purpose_map = search_key.purpose_map.into_iter().map(|(purpose, security_level_map)| {
                 let security_level_map = security_level_map.security_level_map.into_iter().map(|(security_level, key_kind_request_type)| {
                     if security_level > u8::MAX as u32 {
-                        return Err(QueryError::Query(QuerySyntaxError::InvalidKeyParameter("security level out of bounds".to_string())));
+                        return Err(QueryError::DocumentQuery(QuerySyntaxError::InvalidKeyParameter("security level out of bounds".to_string())));
```
Comment (shumkov) [inline packages/rs-drive-abci/src/query/v0/mod.rs:74]: I'm not sure we are querying documents here

File: packages/rs-drive-abci/src/query/v0/mod.rs [hunk 10]
```diff
@@ -115,13 +115,24 @@ impl<C> Platform<C> {
...
+                        "id must be a valid identifier (32 bytes long)".to_string()
+                    )));
                 let response_data = if prove {
                     let proof = check_validation_result_with_data!(self.drive.prove_full_identity(
                         identity_id.into_buffer(),
                         None,
                         &platform_version.drive
                     ));
+
+                    if proof.is_empty() {
```
Comment (shumkov) [inline packages/rs-drive-abci/src/query/v0/mod.rs:130]: I don't think the proof ever be empty. In case of non existence it will be proof of non existence? @QuantumExplorer @lklimek right?

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/rs-drive-abci/src/query/v0/mod.rs (hunk 4 above)
Comment (markin-io) [reply packages/rs-drive-abci/src/query/v0/mod.rs:871]: Will be fixed in the next PR

File: packages/rs-drive-abci/src/abci/handlers.rs (hunk 3 above)
Comment (markin-io) [reply packages/rs-drive-abci/src/abci/handlers.rs]: WIll be fixed in the next PR

File: packages/dapi/lib/externalApis/drive/ErrorCodes.js (hunk 1 above)
Comment (markin-io) [reply packages/dapi/lib/externalApis/drive/ErrorCodes.js]: Fixed

File: packages/rs-drive-abci/src/abci/handlers.rs (hunk 2 above)
Comment (markin-io) [reply packages/rs-drive-abci/src/abci/handlers.rs]: Will be fixed in the next PR

File: packages/rs-drive-abci/src/query/v0/mod.rs (hunk 10 above)
Comment (lklimek) [reply packages/rs-drive-abci/src/query/v0/mod.rs:130]: Agree, we should expect proof of non-existence in this case.

But this check is correct here, empty proof is an error condition and we should return error to the client.

File: packages/rs-drive-abci/src/abci/handlers.rs (hunk 2 above)
Comment (markin-io) [reply packages/rs-drive-abci/src/abci/handlers.rs]: Fixed

File: packages/rs-drive-abci/src/query/v0/mod.rs (hunk 9 above)
Comment (shumkov) [reply packages/rs-drive-abci/src/query/v0/mod.rs:74]: It seems we use query errors for identities and documents. Let's rename DocumentQuery back to Query and refactor it later

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/rs-drive-abci/src/error/handlers.rs (hunk 8 above)
Comment (markin-io) [reply packages/rs-drive-abci/src/error/handlers.rs]: Fixed

File: packages/rs-drive-abci/src/query/v0/mod.rs (hunk 9 above)
Comment (markin-io) [reply packages/rs-drive-abci/src/query/v0/mod.rs]: Fixed

File: packages/rs-drive-abci/src/query/v0/mod.rs (hunk 10 above)
Comment (markin-io) [reply packages/rs-drive-abci/src/query/v0/mod.rs]: Fixed

File: packages/rs-drive-abci/src/abci/handlers.rs [hunk 11]
```diff
@@ -625,23 +616,19 @@ where
                 })
             }
             Err(error) => {
-                let error_data = cbor!({
-                    "message" => "Internal error",
+                let error_data_buffer = platform_value!({
+                    "message": format!("Internal error {}", error.to_string()),
```
[the remark this answers is not in the document]
Comment (markin-io) [reply packages/rs-drive-abci/src/abci/handlers.rs]: Fixed

File: packages/rs-drive-abci/src/abci/handlers.rs [hunk 12]
```diff
@@ -659,12 +646,19 @@ where
         let RequestQuery { data, path, .. } = &request;
 
         let Some(platform_version) = PlatformVersion::get_maybe_current() else {
+            let error_data_buffer = platform_value!({
```
[the remark this answers is not in the document]
Comment (markin-io) [reply packages/rs-drive-abci/src/abci/handlers.rs]: Fixed

File: packages/rs-drive-abci/src/abci/handlers.rs [hunk 13]
```diff
@@ -684,24 +678,24 @@ where
...
-            let error_message = if let Some(error) = error {
-                error.to_string()
-            } else {
-                "Unknown Drive error".to_string()
-            };
+            let handler_error = HandlerError::from(error);
 
-            let mut error_data = Map::new();
-            error_data.insert("message".to_string(), Value::String(error_message));
+            let error_data_buffer = platform_value!({
```
[the remark this answers is not in the document]
Comment (markin-io) [reply packages/rs-drive-abci/src/abci/handlers.rs]: Fixed

**Review by shumkov — DISMISSED** [review]
👍

File: packages/rs-drive-abci/src/abci/handlers.rs [hunk 14]
```diff
@@ -585,20 +583,13 @@ where
                         .serialize_to_bytes_with_platform_version(platform_version)
                         .map_err(|e| ResponseException::from(Error::Protocol(e)))?;
 
-                    let error_data = cbor!({
-                        "data" => {
-                            "serializedError" => consensus_error_bytes
+                    let error_data_buffer = platform_value!({
```
Comment (QuantumExplorer) [inline packages/rs-drive-abci/src/abci/handlers.rs:586]: Why this change?

**Review by QuantumExplorer — CHANGES_REQUESTED** [review]

File: packages/rs-drive-abci/src/query/v0/mod.rs [hunk 15]
```diff
@@ -135,16 +140,19 @@ impl<C> Platform<C> {
...
+                        })
                         .and_then(|identity| identity
-                            .map(|identity| identity
-                                .serialize_consume_to_bytes()
-                                .map_err(QueryError::Protocol))
-                            .transpose()))
-                    .unwrap_or_default();
+                            .serialize_consume_to_bytes()
+                            .map_err(QueryError::Protocol)));
+
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/query/v0/mod.rs:155]: We should respond with not found error then. In case if proofs are requested, we respond with proof of non-existence.

File: packages/rs-drive-abci/src/query/v0/mod.rs [hunk 16]
```diff
@@ -418,16 +448,20 @@ impl<C> Platform<C> {
...
-                        .unwrap())
-                    .map(|contract| {
-                        contract
+                        .unwrap());
+
+                    let data_contract = check_validation_result_with_data!(maybe_data_contract
+                        .ok_or_else(|| {
+                            QueryError::NotFound(format!("data contract {} not found", contract_id))
+                        })
+                        .and_then(|data_contract| data_contract
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/query/v0/mod.rs:457]: That seems correct to me. We respond with not found error code.

File: packages/rs-drive-abci/src/query/v0/mod.rs [hunk 17]
```diff
@@ -783,13 +833,21 @@ impl<C> Platform<C> {
...
-                        .transpose());
+                        .ok_or_else(|| {
+                            QueryError::NotFound(format!(
+                                "identity {} not found",
+                                hex::encode(public_key_hash)
+                            ))
+                        })
+                        .and_then(|identity| identity
+                            .serialize_consume_to_bytes()
+                            .map_err(QueryError::Protocol)));
```
[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/query/v0/mod.rs:844]: That's right. If we requesting multiple items we shouldn't respond with error if one of them not found.

File: packages/rs-drive-abci/src/query/v0/mod.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (markin-io) [reply packages/rs-drive-abci/src/query/v0/mod.rs:457]: Yep, it's similar to singular identity. In case of singular data contract requests we prefer to return NOT FOUND

File: packages/rs-drive-abci/src/query/v0/mod.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (markin-io) [reply packages/rs-drive-abci/src/query/v0/mod.rs:844]: You are confusing it with `identities/by-public-key-hash`. This endpoint supposed to provide one identity, not multiple

[the remark this answers is not in the document]
Comment (shumkov) [reply packages/rs-drive-abci/src/query/v0/mod.rs:844]: Oh, if it's only one then we should use the "not found" error

**Review by QuantumExplorer — DISMISSED** [review]

**Review by shumkov — APPROVED** [review]
